src/home_dir/nix.rs

```diff
@@ -0,0 +1,193 @@
+use std::borrow::BorrowMut;
+use std::cell::RefCell;
+use std::ffi::{CStr, OsString};
+use std::os::unix::ffi::OsStringExt;
+use std::path::PathBuf;
+use std::{env, mem};
+
+use super::{HomeDirError, HomeDirErrorKind};
+
+thread_local! {
+    // Allocate required buffers once per thread instead of each time 'home_dir'
+    // is called so that users who call 'home_dir' multiple times on the same
+    // thread are spared uncessary allocations
+    //
+    // Note: It's prudent here to use heap-allocated buffers instead of stack-
+    // allocated arrays because it's quite difficult (in a cross-platform, 100%
+    // portable way) to ask libc for the maximum possible lengths of usernames
+    // and home directories. There doesn't appear to be a good POSIX standard
+    // for this that all *nix systems adhere to.
+    static BUF0: RefCell<Vec<u8>> = RefCell::new(vec![0; 1024]);
+    static BUF1: RefCell<Vec<u8>> = RefCell::new(vec![0; 1]);
+}
+
+/// Returns the home directory of:
+/// * the current user if `user` is `None` or an empty string, or
+/// * the provided user if `user` is anything else.
+pub(crate) fn home_dir(user: Option<&str>) -> Result<PathBuf, HomeDirError> {
+    if user.is_none() || user == Some("") {
+        if let Some(o) = env::var_os("HOME") {
+            if !o.is_empty() {
+                return Ok(PathBuf::from(o));
+            }
+        }
+    }
+    BUF0.with(move |buf0| {
+        BUF1.with(move |buf1| {
+            let mut buf0 = buf0.borrow_mut();
+            let mut buf1 = buf1.borrow_mut();
+            _home_dir(user, buf0.borrow_mut(), buf1.borrow_mut())
+        })
+    })
+}
+
+fn _home_dir(
+    user: Option<&str>,
+    buf0: &mut Vec<u8>,
+    buf1: &mut Vec<u8>,
```

Comment:
Maybe call these `user_buf` and `path_buf`, or whatever else makes more sense? It seems that they are used for a single purpose here.